### PR TITLE
Add support for mempack

### PR DESCRIPTION
**Pull request by lhchavez**
This allows the creation of git packfiles without needing to write them
to the filesystem.

### CODE REVIEW

File: mempack.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+func (mempack *Mempack) Reset() {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	C.git_mempack_reset(mempack.ptr)
+}
+
+// Free frees the mempack and its resources.
+func (mempack *Mempack) Free() {
+	runtime.SetFinalizer(mempack, nil)
```
Comment (carlosmn) [inline mempack.go]: This still needs to actually free the mempack, not just unset the finalizer.

File: mempack.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	return C.GoBytes(unsafe.Pointer(buf.ptr), C.int(buf.size)), nil
+}
+
+// Reset resets the memory packer by clearing all the queued objects.
+//
+// This assumes that Mempack.Dump has been called before to store all the
+// queued objects into a single packfile.
+func (mempack *Mempack) Reset() {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
```
Comment (carlosmn) [inline mempack.go]: We only need to lock the thread if we're going to deal with errors.

File: mempack.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+	var repoPtr *C.git_repository
+	if repository != nil {
+		repoPtr = repository.ptr
+	}
+
+	ret := C.git_mempack_dump(&buf, repoPtr, mempack.ptr)
+	if ret < 0 {
+		return nil, MakeGitError(ret)
+	}
```
Comment (carlosmn) [inline mempack.go]: We need to keep `repository` alive until here, or the runtime might free it.

File: mempack.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	runtime.SetFinalizer(mempack, (*Mempack).Free)
+
+	ret = C.git_odb_add_backend(odb.ptr, mempack.ptr, C.int(999))
+	if ret < 0 {
+		// Since git_odb_add_alternate() takes ownership of the ODB backend, the
+		// only case in which we free the mempack's memory is if it fails to be
+		// added to the ODB. Mempack.Free() is actually just a no-op.
+		C._go_git_odb_backend_free(mempack.ptr)
+		mempack.Free()
```
Comment (carlosmn) [inline mempack.go]: The free should happen inside `Mempack.Free`.

**Comment by carlosmn** [comment]
Actually since ownership transfers to the repository and the freeing only happens in the constructor, there is no need for a finalizer or `Mempack.Free()` at all.

File: mempack.go (hunk 4 above)
Comment (lhchavez) [reply mempack.go]: Removed

File: mempack.go (hunk 3 above)
Comment (lhchavez) [reply mempack.go]: Done

File: mempack.go (hunk 2 above)
Comment (lhchavez) [reply mempack.go]: Removed

File: mempack.go (hunk 1 above)
Comment (lhchavez) [reply mempack.go]: Removed.

**Comment by carlosmn** [comment]
Thanks, I added an extra keepalive that was missing (these are _so_ annoying).

**Pull request merged**